libp2p/host.ss

```diff
@@ -0,0 +1,78 @@
+;;; -*- Gerbil -*-
+;;; libp2p host implementation
+
+(import :std/io
+        :std/misc/timeout
+        :std/format
+        :std/sugar
+        :std/crypto
+        :std/text/base58
+        :std/logger
+        "utils"
+        "multiaddress")
+
+;; TODO This should derive the peer-id from the key-pair
+(def (make-peer-id key)
+  (base58-encode key))
+
+(defstruct peer-info (peer-id multi-address public-key))
+
+(defstruct p2p-host (id key multi-address listener peer-store handlers connections)
+  constructor: init!)
+
+(defmethod {init! p2p-host}
+  (lambda (self (key (keygen/ed25519)) (multi-addr "ip4/127.0.0.1/tcp/4444"))
+    (set! (@ self multi-address) (multi-address multi-addr))
+    (set! (@ self peer-store) (make-hash-table))
+    (set! (@ self handlers) (make-hash-table))
+    (set! (@ self connections) (make-hash-table))
+    ;; TODO Make this a proper implementation
+    (set! (@ self id) (make-peer-id key))))
+
+(def (tcp-listener address (proc 123))
+  (let* ((addr (get-tcp-addr address))
+         (sock (tcp-listen addr)))
+    (while #t
+      (try
+       (let (cli (ServerSocket-accept sock))
+         (when cli
+           (println  (format "~a" cli))
+           (println (format "Accepted connection from ~a" (StreamSocket-peer-address cli)))
+           ;;(spawn proc cli)
+           ))
+       (catch (e)
+         (errorf "Error accepting connection: ~a" e))))))
+
```

Comment:
use the dots, luke!

```
(using (sock (tcp-listen addr) :- Serverocket)
   (while #t
      (using (cli (sock.accept) :- StreamSoscket)
         ...
         ... cli.peer-address ...)))
```

Reply:
Also of note, you need to allow an exception to exit the loop if the socket is closed by another thread -- in fact you should make the socket closable by putting it in the host -- so that you can cleanly exit.